drivers/pio/sunxi-pio.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <bitmap.h>
+#include <dm.h>
+#include <error.h>
+#include <mmio.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <util.h>
+#include <drivers/clock.h>
+#include <drivers/irqchip.h>
+#include <drivers/pio.h>
+#include <drivers/pio/sunxi-pio.h>
+
+#define MAX_PORTS      8
+
+#define PIN_INDEX(pin) ((pin) & BITMASK(0, 5))
+#define PIN_PORT(pin)  ((pin) >> 5)
+
+#define PIN_MODE(mode) ((mode) & BITMASK(0, 3))
+
+#define CONFIG_REG(pin) \
+	(0x0000 + PIN_PORT(pin) * 0x24 + (PIN_INDEX(pin) >> 3) * 4)
+#define DATA_REG(pin) \
+	(0x0010 + PIN_PORT(pin) * 0x24)
+
+#define CONFIG_OFFSET(pin)    ((PIN_INDEX(pin) & BITMASK(0, 3)) * 4)
+
+#define INT_CONTROL_REG(port) (0x0210 + (port) * 0x20)
+#define INT_STATUS_REG(port)  (0x0214 + (port) * 0x20)
+
+static bool
+sunxi_pio_read_pin(struct device *dev, uint8_t pin)
+{
+	return mmio_read32(dev->regs + DATA_REG(pin)) & BIT(PIN_INDEX(pin));
+}
+
+static int
+sunxi_pio_set_mode(struct device *dev, uint8_t pin, uint8_t mode)
+{
+	/* Verify port exists. */
+	if (!bitmap_get(dev->drvdata, PIN_PORT(pin)))
```

Comment:
This should be something like
```
if (!(dev->drvdata & BIT(PIN_PORT(pin))))
```
Although `bitmap_get` takes a `uintptr_t` as a first parameter (because it passes it along to `mmio_read32`), it's treated as a pointer. So the code here as written treats the *contents* of `dev->drvdata` as the *address* of the bitmap. You could write it as
```
if (!bitmap_get((uintptr_t)&dev->drvdata, PIN_PORT(pin)))
```
but this is more verbose and does some unnecessary math (since you know there will never be more than one word's work of ports).

Reply:
Fixed.